README.md
# Javascript Storage Manager

A library that makes the [WebStorage](https://en.wikipedia.org/wiki/Web_storage) easy to use: `localStorage`, `sessionStorage`  and `Cookies` (`sessionStorage` and `Cookies` are in progress: [#2](/../../issues/2), [#3](/../../issues/3)).

![npm](https://img.shields.io/npm/v/js-storage-manager.svg) ![NPM](https://img.shields.io/npm/l/js-storage-manager.svg) ![npm bundle size](https://img.shields.io/bundlephobia/min/js-storage-manager.svg)



&nbsp;
&nbsp;

## Install

### bower

```bash
$ bower install js-storage-manager
```

### npm

```bash
$ npm install js-storage-manager
```

### Git

```bash
$ git clone https://github.com/bjoern-hempel/js-storage-manager.git
```

or

```bash
$ git clone git@github.com:bjoern-hempel/js-storage-manager.git
```



&nbsp;
&nbsp;

## Getting started

### bower

```bash
$ mkdir webproject && cd webproject
$ bower install js-storage-manager
$ vi index.html
```

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>A simple js-storage-manager example</title>
    <script src="bower_components/js-storage-manager/dist/storage-manager.min.js"></script>
  </head>
  <body>
    <script>
      var sm = new StorageManager('namespace');

      sm.set('data', [{id: 123, name: 'Name 1'}, {id: 123, name: 'Name 2'}]);

      document.write(JSON.stringify(sm.get('data')));
    </script>
  </body>
</html>
```

### npm

#### The direct way (the old javascript way)

```bash
$ mkdir webproject && cd webproject
$ npm install js-storage-manager
$ vi index.html
```

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>A simple js-storage-manager example</title>
    <script src="node_modules/js-storage-manager/dist/storage-manager.min.js"></script>
  </head>
  <body>
    <script>
      var sm = new StorageManager('namespace');

      sm.set('data', [{id: 123, name: 'Name 1'}, {id: 123, name: 'Name 2'}]);

      document.write(JSON.stringify(sm.get('data')));
    </script>
  </body>
</html>
```

#### The webpack way (the modern javascript way)

If you are interested to use this library on the modern javascript way, see [here](md/the_webpack_way.md).

### Git

```bash
$ mkdir webproject && cd webproject
$ mkdir vendor && cd vendor
$ git clone https://github.com/bjoern-hempel/js-storage-manager.git
$ cd ..
$ vi index.html
```

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>A simple js-storage-manager example</title>
    <script src="vendor/js-storage-manager/dist/storage-manager.min.js"></script>
  </head>
  <body>
    <script>
      var sm = new StorageManager('namespace');

      sm.set('data', [{id: 123, name: 'Name 1'}, {id: 123, name: 'Name 2'}]);

      document.write(JSON.stringify(sm.get('data')));
    </script>
  </body>
</html>
```



&nbsp;
&nbsp;

## The next steps

### How to use the StorageManager in the easiest way

```javascript
var sm = new StorageManager('namespace');
var data_initial = [{id: 1, name: 'Name 1'}, {id: 2, name: 'Name 2'}];

/* save data_initial to localStorage.storage.namespace.data */
sm.set('data', data_initial);

var data_from_web_storage = sm.get('data');
```

### How to use multiple namespaces

```javascript
var namespace_1 = 'namespace1';
var namespace_2 = 'namespace2';

var sm_1 = new StorageManager(namespace_1);
var sm_2 = new StorageManager(namespace_2);

var data_initial_1 = [{id: 1, name: 'Name 1'}, {id: 2, name: 'Name 2'}];
var data_initial_2 = [{id: 3, name: 'Name 3'}, {id: 4, name: 'Name 4'}];

/* save data_initial to localStorage.storage.namespace1.data */
sm.set('data', data_initial_1);

/* save data_initial to localStorage.storage.namespace2.data */
sm.set('data', data_initial_2);

var data_from_web_storage_1 = sm_1.get('data');
var data_from_web_storage_2 = sm_2.get('data');
```

### How to manage the storage yourself

```javascript
var sm = new StorageManager('namespace');
var data_initial = [{id: 1, name: 'Name 1'}, {id: 2, name: 'Name 2'}];

/* Get the storage data object. */
var storage = sm.getStorage();

/* Do something with the data object. */
storage.data = data_initial;

/* Manually save the data object in WebStorage. */
sm.setStorage(storage)
```

### How to let the StorageManager automatically save changes to the storage data object in web storage

```javascript
var sm = new StorageManager('namespace');
var data_initial = [{id: 1, name: 'Name 1'}, {id: 2, name: 'Name 2'}];
var observable = true; // <- important

/* The returned storage data object is now of type "Proxy". */
var storage = sm.getStorage(observable);

/* Do something with the data object. */
storage.data = data_initial;
/* sm.setStorage(storage) is no longer needed. Changes are automatically saved in WebStorage. */
```

### How to create and use a queue list

```javascript
var sm = new StorageManager('namespace');

var queue_data_1 = {id: 1, name: 'Name 1'};
var queue_data_2 = {id: 2, name: 'Name 2'};

/* Initialize the queue (optionally) */
var qm = sm.initQueue();

/* Add records to the queue. The queue namespace used is 'queue'.
 * Attention. If LocalStorage is used, this value is added again and again. Use the Reset parameter within
 * initQueue to clear the persistent memory before.
 */
qm.push(queue_data_1);
qm.push(queue_data_2);

/* Get the number of queue items. */
var number_of_queue_items = qm.getNumber();

/* Read the entire queue */
var queue = qm.getAll();

/* Get the next queue item (FIFO) */
var next_queue_item = qm.getNext();

/* Get the next queue entry and delete it. */
var next_queue_item = qm.deleteNext();
```

### How to use your own queue namespace or multiple instances of a queue within a namespace

```javascript
var sm = new StorageManager('namespace');

var queue_data_1 = {id: 1, name: 'Name 1'};
var queue_data_2 = {id: 2, name: 'Name 2'};

/* Initialize the queue (optionally) */
var qm = sm.initQueue('my_queue', true);
```

### How to get the LocalStorage area completely managed by the StorageManager

```javascript
var sm = new StorageManager('namespace');

/* Returns the LocalStorage area as a ready-parsed object. */
var local_storage_managed_by_sm = sm.getLocalStorage();
```



&nbsp;
&nbsp;

## Maintenance

1. Checkout the repository

```bash
$ git clone git@github.com:bjoern-hempel/js-storage-manager.git && cd js-storage-manager
$ npm install
```

2. Extend, fix bugs in classes below `/src` folder. 
3. Write more tests below the `/test` folder.
4. Run the tests.

```bash
$ npm test
```

or

```bash
$ npm run test:unit
```

5. Build the `/dist` files

```bash
$ npm run build
```

6. Change the version number

```bash
$ vi package.json
```

```text
...
"version": "0.0.14",
...
```

7. Commit your changes

```bash
$ git add [file1] [file2] [etc.]
$ git commit -m "my bugfixes" .
$ git push
```

8. Tag your version

```bash
$ git tag v0.0.14
$ git push origin v0.0.14
```

9. Create Release

If necessary:

```bash
$ sudo npm install github-release-notes -g
```

Then:

```bash
$ gren release
```

**Important:** You need a valid Github token to access the API. You can get your own here: https://github.com/settings/tokens

Adapt the changelog text to github if necessary: [changelog](https://github.com/bjoern-hempel/js-storage-manager/releases). Show all commits:

```bash
$ git log --oneline --decorate
```

10. Update [CHANGELOG.md](/CHANGELOG.md)

```bash
$ gren changelog --override
$ git commit -m "Change changelog" .
$ git push
```

11. Publish to npm

If necessary:

```bash
$ npm login
```

Then:

```bash
$ npm publish
```



&nbsp;
&nbsp;

## A. Authors

* Björn Hempel <bjoern@hempel.li> - _Initial work_ - [https://github.com/bjoern-hempel](https://github.com/bjoern-hempel)



&nbsp;
&nbsp;

## B. Changelog

Changes are tracked as [GitHub releases](https://github.com/bjoern-hempel/js-storage-manager/releases) or in reverse order [here](/CHANGELOG.md).



&nbsp;
&nbsp;

## C. License

This library is licensed under the MIT License - see the [LICENSE.md](/LICENSE.md) file for details


CHANGELOG.md
# Changelog

## v0.0.20 (12/04/2019)
#### New:

* Fix the main file of the package for use as npm module.
* Some documentation changes.
---

## v0.0.19 (11/04/2019)
#### New:

* Dist generated js files are now backward compatible (es2015)
* The package-lock.json is now also stored in the project
---

## v0.0.18 (18/03/2019)
#### New:

* [**closed**] Build the queue management area into its own object. [#1](https://github.com/bjoern-hempel/js-storage-manager/issues/1)
* Tidy up the documentation.

---

## v0.0.17 (18/03/2019)
#### New:

* Add a queue management class into its own object (See #1)

---

## v0.0.16 (17/03/2019)
#### New:

* Adding "how to" items in the description
---

## v0.0.15 (17/03/2019)
#### New:

* Add CHANGELOG.md to repository
---

## v0.0.1 (17/03/2019)
#### New:

* This is the moment when this repository sees the light of day and already works quite well.
---

## v0.0.2 (17/03/2019)
#### New:

* Add some more descriptions to StorageManager class

#### Fixes:

* The "auto-save" function now works again ([`Proxy`](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Proxy))
---

## v0.0.3 (17/03/2019)
#### New:

* Add "auto-save" for deep objects ([`Proxy`](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Proxy))
* Add the queue functions
---

## v0.0.4 (17/03/2019)
#### New:

* Add the index.js and some customizations to [package.json](https://docs.npmjs.com/files/package.json)
---

## v0.0.5 (17/03/2019)
#### New:

* Add the first unit tests
---

## v0.0.6 (17/03/2019)
#### New:

* Add some more unit tests
---

## v0.0.7 (17/03/2019)
#### New:

* Adds directly usable and precompiled packages within the `dist` directory
* Adds `npm run build` to simply generate `dist` packages
---

## v0.0.8 (17/03/2019)
#### New:

* Just some refactoring
---

## v0.0.9 (17/03/2019)
#### New:

* Just some refactoring
---

## v0.0.10 (17/03/2019)
#### Fixes:

* Storage objects are now directly accessible when using dist packages
---

## v0.0.11 (17/03/2019)
#### New:

* Renames the `Storage` object into a less interchangeable name: `StorageManager`
---

## v0.0.12 (17/03/2019)
#### New

* Tidy up the [package.json](https://docs.npmjs.com/files/package.json)
* Publish the latest `dist` packages
---

## v0.0.13 (17/03/2019)
#### New:

* First version with [bower](https://bower.io) support
---

## v0.0.14 (17/03/2019)
#### New:

* Some fine tunings for [bower](https://bower.io)

test/storage.spec.js
/**
 * Simple storage test runner
 *
 * run with:
 * user$ npm run test:unit
 */

/* global describe, beforeEach, afterEach, localStorage, it */
import 'mock-local-storage'

let Storage = require('../src/storage')
let assert = require('assert')

const STORAGE_NAMESPACE = 'storageNamespace'

describe('export class Storage {}', () => {
	beforeEach(() => {
		/* Mock the localStorage */
		global.window = {}
		window.localStorage = global.localStorage
	})

	afterEach(() => {
		/* Clear the localStorage */
		localStorage.clear();
		localStorage.itemInsertionCallback = null;
	})

	/**
	 * Test some basic functions without the localStorage.
	 */
	describe('Basic function tests.', () => {
		it('Test Storage.get default value: null.', () => {
			/* Arrange */
			let dataExpected = null

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			let returnData = s.get('data')

			/* Assert */
			assert.deepStrictEqual(returnData, dataExpected)
		})

		it('Test Storage.get given default value without given values before: [1, 2, 3]', () => {
			/* Arrange */
			let dataExpected = [1, 2, 3]

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			let dataReturn = s.get('data', dataExpected)

			/* Assert */
			assert.deepStrictEqual(dataReturn, dataExpected)
		})

		it('Test Storage.get given default value with given values before: [1, 2, 3]', () => {
			/* Arrange */
			let data1 = [1, 2, 3]
			let data2 = [4, 5, 6]
			let dataExpected = data1

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.set('data', data1)
			let dataReturn = s.get('data', data2)

			/* Assert */
			assert.deepStrictEqual(dataReturn, dataExpected)
		})

		it('Test Storage.getStorage function.', () => {
			/* Arrange */
			let data = []
			let dataExpected = { 'data': data }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.set('data', data)
			let dataReturn = s.getStorage()

			/* Assert */
			assert.deepStrictEqual(dataReturn, dataExpected)
		})
	})

	/**
	 * Test the localStorage without the observable magic methods.
	 */
	describe('Empty localStorage (non observable).', () => {
		it('Test localStorage which should be empty first.', () => {
			/* Arrange */
			let localStorageExpected = {}
			localStorageExpected[STORAGE_NAMESPACE] = {}

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.initStorage()

			/* Assert */
			assert.deepStrictEqual(s.getLocalStorage(), localStorageExpected)
		})

		it('Test localStorage with Storage.set function.', () => {
			/* Arrange */
			let localStorageExpected = {}
			let data = [1, 2, 3]
			localStorageExpected[STORAGE_NAMESPACE] = { data: data }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.set('data', data)

			/* Assert */
			assert.deepStrictEqual(s.getLocalStorage(), localStorageExpected)
		})

		it('Test Storage.initQueue (empty queue).', () => {
			/* Arrange */
			let localStorageExpected = {}
			localStorageExpected[STORAGE_NAMESPACE] = { queue: [] }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.initQueue()

			/* Assert */
			assert.deepStrictEqual(s.getLocalStorage(), localStorageExpected)
		})

		it('Test Storage.addQueue (one entry).', () => {
			/* Arrange */
			let localStorageExpected = {}
			let data = { 'id': 123, 'name': 'Name' }
			localStorageExpected[STORAGE_NAMESPACE] = { queue: [data] }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.pushQueue(data)

			/* Assert */
			assert.deepStrictEqual(s.getLocalStorage(), localStorageExpected)
		})

		it('Test Storage.addQueue (two entries).', () => {
			/* Arrange */
			let localStorageExpected = {}
			let data = { 'id': 123, 'name': 'Name' }
			localStorageExpected[STORAGE_NAMESPACE] = { queue: [data, data] }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.pushQueue(data)
			s.pushQueue(data)

			/* Assert */
			assert.deepStrictEqual(s.getLocalStorage(), localStorageExpected)
		})

		it('Test Storage.getNumberOfQueuesItems (three entries).', () => {
			/* Arrange */
			let data = { 'id': 123, 'name': 'Name' }
			let numberExpected = 3

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.pushQueue(data)
			s.pushQueue(data)
			s.pushQueue(data)

			/* Assert */
			assert.equal(s.getNumberOfQueuesItems(), numberExpected)
		})

		it('Test Storage.getQueue (two entries).', () => {
			/* Arrange */
			let data = { 'id': 123, 'name': 'Name' }
			let queueExpected = [data, data]

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.pushQueue(data)
			s.pushQueue(data)

			/* Assert */
			assert.deepStrictEqual(s.getQueue(), queueExpected)
		})

		it('Test Storage.getNextQueueEntry.', () => {
			/* Arrange */
			let data1 = { 'id': 1, 'name': 'Name 1' }
			let data2 = { 'id': 2, 'name': 'Name 2' }
			let data3 = { 'id': 3, 'name': 'Name 3' }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.pushQueue(data1)
			s.pushQueue(data2)
			s.pushQueue(data3)

			/* Assert */
			assert.deepStrictEqual(s.getNextQueueItem(), data1)
		})

		it('Test Storage.deleteNextQueueEntry (get deleted entry).', () => {
			/* Arrange */
			let data1 = { 'id': 1, 'name': 'Name 1' }
			let data2 = { 'id': 2, 'name': 'Name 2' }
			let data3 = { 'id': 3, 'name': 'Name 3' }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.pushQueue(data1)
			s.pushQueue(data2)
			s.pushQueue(data3)

			/* Assert */
			assert.deepStrictEqual(s.deleteNextQueueItem(), data1)
		})

		it('Test Storage.deleteNextQueueEntry (check new queue list).', () => {
			/* Arrange */
			let data1 = { 'id': 1, 'name': 'Name 1' }
			let data2 = { 'id': 2, 'name': 'Name 2' }
			let data3 = { 'id': 3, 'name': 'Name 3' }
			let queueExpected = [data2, data3]

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			s.pushQueue(data1)
			s.pushQueue(data2)
			s.pushQueue(data3)
			s.deleteNextQueueItem()

			/* Assert */
			assert.deepStrictEqual(s.getQueue(), queueExpected)
		})
	})

	/**
	 * Test the localStorage with the observable magic methods.
	 */
	describe('Empty localStorage (observable).', () => {
		it('Test auto-save to localStorage (Simple).', () => {
			/* Arrange */
			let localStorageExpected = {}
			let data = { 'id': 1, 'name': 'Name 1', 'queue': ['a', 'b', 'c'] }
			localStorageExpected[STORAGE_NAMESPACE] = { data: data }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			let observable = true
			let storage = s.getStorage(observable)
			storage.data = data /* auto save to localStorage */

			/* Assert */
			assert.deepStrictEqual(s.getLocalStorage(), localStorageExpected)
		})

		it('Test auto-save to localStorage (Deep).', () => {
			/* Arrange */
			let localStorageExpected = {}
			let data = { 'id': 1, 'name': 'Name 1', 'queue': ['a', 'b', 'c'] }
			let dataExpected = { 'id': 2, 'name': 'Name 1', 'queue': ['a', 'd', 'c'] }
			localStorageExpected[STORAGE_NAMESPACE] = { data: dataExpected }

			/* Act */
			let s = new Storage(STORAGE_NAMESPACE)
			let observable = true
			let storage = s.getStorage(observable)
			storage.data = data /* auto save to localStorage */
			storage.data.id = 2 /* auto save to localStorage */
			storage.data.queue[1] = 'd' /* auto save to localStorage */

			/* Assert */
			assert.deepStrictEqual(s.getLocalStorage(), localStorageExpected)
		})
	})
})



src/storage.js
/**
 * Class to save data to localStorage.
 *
 *
 * Usage (without observable flag):
 * --------------------------------
 *
 * // Create the "test" storage (localStore.storage.test)
 * let ts = new Storage('test')
 *
 * // Get the entire "test" storage
 * let observable = false
 * testStorage = ts.getStorage(observable)
 *
 * // Save the entire storage to localStorage (localStorage.storage.test)
 * let data = [{'id': 123, 'name': 'Björn Hempel'}]
 * testStorage.data = data
 * ts.setStorage(testStorage)
 *
 * // Save a part of the storage to localStorage (localStorage.storage.test.data2)
 * let data2 = [{'id': 123, 'name': 'Björn Hempel'}]
 * ts.set('data2', data2)
 * console.log(testStorage.data2) // testStorage.data2 = data2 is not needed
 *
 * // Get a part of the storage (localStorage.storage.test.data)
 * let data = ts.get('data')
 *
 * // output the entire "test" storage (localStorage.storage.test.data && localStorage.storage.test.data2)
 * console.log(testStorage)
 *
 *
 * Usage (with observable flag - Proxy):
 * -------------------------------------
 *
 * // Create the "test" storage (localStore.storage.test)
 * let ts = new Storage('test')
 *
 * // Get the entire "test" storage as observable object
 * let observable = true
 * testStorage = ts.getStorage(observable)
 *
 * // Now all changes on testStorage are automatically saved into the localStorage
 * let data = [{'id': 123, 'name': 'Björn Hempel'}]
 * testStorage['data'] = data // auto-save to localstorage
 * // saved in localStorage: ls.storage.test.data[].id = 123, ls.storage.test.data[].name = 'Björn Hempel'
 * // same as: ts.set('data', [{'id': 123, 'name': 'Björn Hempel'}]) -> wenn observable = false
 *
 * // output the entire "test" storage (localStorage.storage.test.data)
 * console.log(testStorage)
 *
 *
 * @author  Björn Hempel <bjoern@hempel.li>
 * @version 1.0 (2019-03-15)
 */
"use strict"

const LS_STORAGE_KEY = 'storage'

class Storage {
	constructor (area) {
		this.ls = window.localStorage
		this.area = area

		this.initLocalStorage()
	}

	/**
	 * Initializes the entire "LocalStorage" storage.
	 *
	 * @param defaultValue
	 */
	initLocalStorage (defaultValue = {}) {
		if (this.ls.getItem(LS_STORAGE_KEY) === null) {
			this.ls.setItem(LS_STORAGE_KEY, JSON.stringify(defaultValue))
		}
	}

	/**
	 * Returns the entire content of the storage (in parsed state).
	 *
	 * @returns {any}
	 */
	getLocalStorage () {
		this.initLocalStorage()

		return JSON.parse(this.ls.getItem(LS_STORAGE_KEY))
	}

	/**
	 * Sets the entire storage within the LocalStorage.
	 *
	 * @param value
	 */
	setLocalStorage (value) {
		this.initLocalStorage()

		this.ls.setItem(LS_STORAGE_KEY, JSON.stringify(value))
	}

	/**
	 * Initialize an area of the storage.
	 *
	 * @param defaultValue
	 * @returns {*}
	 */
	initStorage (defaultValue = {}) {
		let localStorage = this.getLocalStorage()

		if (!localStorage.hasOwnProperty(this.area)) {
			this.setStorage(defaultValue)
		}

		if (typeof this.storage === 'undefined') {
			this.storage = localStorage[this.area]
		}
	}

	/**
	 * Returns an area of the storage.
	 *
	 * @returns {*}
	 */
	getStorage (observable = false) {
		this.initStorage()

		if (observable) {
			return this.observe(
				this.storage,
				(property, value) => {
					let data = this.storage

					let properties = property.split('.')
					let lastProperty = properties.pop()

					/* Iterate through given properties. */
					properties.forEach(function (key) {
						data = data[key]
					})

					data[lastProperty] = value

					this.setStorage(this.storage)
				}
			)
		}

		return this.storage
	}

	/**
	 * Sets an area of the storage with value.
	 *
	 * @param storage
	 */
	setStorage (storage) {
		let localStorage = this.getLocalStorage()

		localStorage[this.area] = storage

		this.storage = storage

		this.setLocalStorage(localStorage)
	}

	/**
	 * Gets a key from the storage (this.area).
	 *
	 * @param key
	 * @returns {*}
	 */
	get (key, defaultValue = null, observable = false) {
		let storageArea = this.getStorage(observable)

		if (!storageArea.hasOwnProperty(key)) {
			return defaultValue
		}

		return storageArea[key]
	}

	/**
	 * Sets a key within the storage (this.area).
	 *
	 * @param key
	 * @param value
	 */
	set (key, value) {
		let storageArea = this.getStorage()

		storageArea[key] = value

		this.setStorage(storageArea)
	}

	/**
	 * Initialize a queue.
	 *
	 * @param key
	 */
	initQueue (key = 'queue') {
		let values = this.get(key, [])

		if (!Array.isArray(values)) {
			console.error('initQueue: The given values must be an array!')
			return
		}

		this.set(key, values)
	}

	/**
	 * Returns all entries of the queue.
	 *
	 * @param key
	 */
	getQueue (key = 'queue') {
		this.initQueue(key)

		let values = this.get(key, [])

		if (!Array.isArray(values)) {
			return []
		}

		return values
	}

	/**
	 * Adds the given element to the queue.
	 *
	 * @param value
	 * @param key
	 */
	pushQueue (value, key = 'queue') {
		this.initQueue(key)

		let values = this.get(key, [])

		if (!Array.isArray(values)) {
			console.error('pushQueue: The given values must be an array!')
			return
		}

		values.push(value)

		this.set(key, values)
	}

	/**
	 * Gets the number of queue entries.
	 *
	 * @param value
	 * @param key
	 */
	getNumberOfQueuesItems (key = 'queue') {
		return this.getQueue(key).length
	}

	/**
	 * Gets the next queue entry (FIFO).
	 *
	 * @param value
	 * @param key
	 */
	getNextQueueItem (key = 'queue') {
		this.initQueue(key)

		let values = this.get(key, [])

		if (!Array.isArray(values)) {
			return null
		}

		if (this.getNumberOfQueuesItems(key) <= 0) {
			return null
		}

		return values[0]
	}

	/**
	 * Deletes the next queue entry and returns it (FIFO).
	 *
	 * @param value
	 * @param key
	 */
	deleteNextQueueItem (key = 'queue') {
		this.initQueue(key)

		let values = this.get(key, [])

		if (!Array.isArray(values)) {
			return null
		}

		if (this.getNumberOfQueuesItems(key) <= 0) {
			return null
		}

		return values.shift()
	}

	/**
	 * Adds magic getters and setter methods to the given object to be able to react to changes with a
	 * callback function.
	 *
	 * @param obj
	 * @param callback
	 */
	observe (obj, callback) {
		function buildProxy (prefix, obj) {
			let changeHandler = {
				get: (target, property) => {
					const out = target[property]

					if (out instanceof Object) {
						return buildProxy(prefix + property + '.', out)
					}

					return out
				},
				set: (target, property, value) => {
					callback(prefix + property, value)

					target[property] = value

					return true
				}
			}

			return new Proxy(obj, changeHandler)
		}

		return buildProxy('', obj)
	}
}

module.exports = Storage



src/storage-manager.js
/**
 * Class to save data to localStorage.
 *
 *
 * Usage (without observable flag):
 * --------------------------------
 *
 * // Create the "test" storage (localStore.storage.test)
 * let ts = new Storage('test')
 *
 * // Get the entire "test" storage
 * let observable = false
 * testStorage = ts.getStorage(observable)
 *
 * // Save the entire storage to localStorage (localStorage.storage.test)
 * let data = [{'id': 123, 'name': 'Björn Hempel'}]
 * testStorage.data = data
 * ts.setStorage(testStorage)
 *
 * // Save a part of the storage to localStorage (localStorage.storage.test.data2)
 * let data2 = [{'id': 123, 'name': 'Björn Hempel'}]
 * ts.set('data2', data2)
 * console.log(testStorage.data2) // testStorage.data2 = data2 is not needed
 *
 * // Get a part of the storage (localStorage.storage.test.data)
 * let data = ts.get('data')
 *
 * // output the entire "test" storage (localStorage.storage.test.data && localStorage.storage.test.data2)
 * console.log(testStorage)
 *
 *
 * Usage (with observable flag - Proxy):
 * -------------------------------------
 *
 * // Create the "test" storage (localStore.storage.test)
 * let ts = new Storage('test')
 *
 * // Get the entire "test" storage as observable object
 * let observable = true
 * testStorage = ts.getStorage(observable)
 *
 * // Now all changes on testStorage are automatically saved into the localStorage
 * let data = [{'id': 123, 'name': 'Björn Hempel'}]
 * testStorage['data'] = data // auto-save to localstorage
 * // saved in localStorage: ls.storage.test.data[].id = 123, ls.storage.test.data[].name = 'Björn Hempel'
 * // same as: ts.set('data', [{'id': 123, 'name': 'Björn Hempel'}]) -> wenn observable = false
 *
 * // output the entire "test" storage (localStorage.storage.test.data)
 * console.log(testStorage)
 *
 *
 * @author  Björn Hempel <bjoern@hempel.li>
 * @version 1.0 (2019-03-15)
 */
"use strict"

const LS_STORAGE_KEY = 'storage'

class StorageManager {
	constructor (area) {
		this.ls = window.localStorage
		this.area = area

		this.initLocalStorage()
	}

	/**
	 * Initializes the entire "LocalStorage" storage.
	 *
	 * @param defaultValue
	 */
	initLocalStorage (defaultValue = {}) {
		if (this.ls.getItem(LS_STORAGE_KEY) === null) {
			this.ls.setItem(LS_STORAGE_KEY, JSON.stringify(defaultValue))
		}
	}

	/**
	 * Returns the entire content of the storage (in parsed state).
	 *
	 * @returns {any}
	 */
	getLocalStorage () {
		this.initLocalStorage()

		return JSON.parse(this.ls.getItem(LS_STORAGE_KEY))
	}

	/**
	 * Sets the entire storage within the LocalStorage.
	 *
	 * @param value
	 */
	setLocalStorage (value) {
		this.initLocalStorage()

		this.ls.setItem(LS_STORAGE_KEY, JSON.stringify(value))
	}

	/**
	 * Initialize an area of the storage.
	 *
	 * @param defaultValue
	 * @returns {*}
	 */
	initStorage (defaultValue = {}) {
		let localStorage = this.getLocalStorage()

		if (!localStorage.hasOwnProperty(this.area)) {
			this.setStorage(defaultValue)
		}

		if (typeof this.storage === 'undefined') {
			this.storage = localStorage[this.area]
		}
	}

	/**
	 * Returns an area of the storage.
	 *
	 * @returns {*}
	 */
	getStorage (observable = false) {
		this.initStorage()

		if (observable) {
			return this.observe(
				this.storage,
				(property, value) => {
					let data = this.storage

					let properties = property.split('.')
					let lastProperty = properties.pop()

					/* Iterate through given properties. */
					properties.forEach(function (key) {
						data = data[key]
					})

					data[lastProperty] = value

					this.setStorage(this.storage)
				}
			)
		}

		return this.storage
	}

	/**
	 * Sets an area of the storage with value.
	 *
	 * @param storage
	 */
	setStorage (storage) {
		let localStorage = this.getLocalStorage()

		localStorage[this.area] = storage

		this.storage = storage

		this.setLocalStorage(localStorage)
	}

	/**
	 * Gets a key from the storage (this.area).
	 *
	 * @param key
	 * @returns {*}
	 */
	get (key, defaultValue = null, observable = false) {
		let storageArea = this.getStorage(observable)

		if (!storageArea.hasOwnProperty(key)) {
			return defaultValue
		}

		return storageArea[key]
	}

	/**
	 * Sets a key within the storage (this.area).
	 *
	 * @param key
	 * @param value
	 */
	set (key, value) {
		let storageArea = this.getStorage()

		storageArea[key] = value

		this.setStorage(storageArea)
	}

	/**
	 * Initialize a queue.
	 *
	 * @param key
	 */
	initQueue (key = 'queue', reset = false) {
		return new QueueManager(this, key, reset)
	}

	/**
	 * Adds magic getters and setter methods to the given object to be able to react to changes with a
	 * callback function.
	 *
	 * @param obj
	 * @param callback
	 */
	observe (obj, callback) {
		function buildProxy (prefix, obj) {
			let changeHandler = {
				get: (target, property) => {
					const out = target[property]

					if (out instanceof Object) {
						return buildProxy(prefix + property + '.', out)
					}

					return out
				},
				set: (target, property, value) => {
					callback(prefix + property, value)

					target[property] = value

					return true
				}
			}

			return new Proxy(obj, changeHandler)
		}

		return buildProxy('', obj)
	}
}

class QueueManager {
	constructor (sm, key = 'queue', reset = false) {
		this.sm = sm
		this.key = key

		this.init(reset)
	}

	/**
	 * Initialize the queue.
	 */
	init (reset = false) {
		if (reset) {
			this.sm.set(this.key, [])
		}

		let values = this.sm.get(this.key, [])

		if (!Array.isArray(values)) {
			console.error('initQueue: The given values must be an array!')
			return
		}

		this.sm.set(this.key, values)
	}

	/**
	 * Returns all entries of the queue.
	 *
	 * @param key
	 */
	getAll () {
		this.init()

		let values = this.sm.get(this.key, [])

		if (!Array.isArray(values)) {
			return []
		}

		return values
	}

	/**
	 * Adds the given element to the queue.
	 *
	 * @param value
	 * @param key
	 */
	push (value) {
		this.init()

		let values = this.sm.get(this.key, [])

		if (!Array.isArray(values)) {
			console.error('pushQueue: The given values must be an array!')
			return
		}

		values.push(value)

		this.sm.set(this.key, values)
	}

	/**
	 * Gets the number of queue entries.
	 *
	 * @param value
	 * @param key
	 */
	getNumber () {
		return this.getAll().length
	}

	/**
	 * Gets the next queue entry (FIFO).
	 *
	 * @param value
	 * @param key
	 */
	getNext (key = 'queue') {
		this.init()

		let values = this.sm.get(this.key, [])

		if (!Array.isArray(values)) {
			return null
		}

		if (this.getNumber() <= 0) {
			return null
		}

		return values[0]
	}

	/**
	 * Deletes the next queue entry and returns it (FIFO).
	 *
	 * @param value
	 * @param key
	 */
	deleteNext (key = 'queue') {
		this.init()

		let values = this.sm.get(key, [])

		if (!Array.isArray(values)) {
			return null
		}

		if (this.getNumber() <= 0) {
			return null
		}

		return values.shift()
	}
}

module.exports = StorageManager
